### PR TITLE
Adds Database.Orville.PostgreSQL.Plan for planning queries

### PR DESCRIPTION
**Pull request by qxjit**
This adds a new 'Plan' type that is meant to be the replacement for
'Popper' in the eventual libpq based Hackage release. Rather than using
'Arrow' as the primary interface for chaining queries together, 'Plan'
provides 'bind' function that is quite similar the 'Monad' bind function
'>>=', though does not follow the exact same type signature. Although
'Plan' cannot implement 'Monad', having 'bind' function will enable the
use of GHC 9 'QualifiedDo' extension.

A future PR will provide the interface required for the extension along
with examples. This PR is introduces the type so that we can start using
it before GHC9 is more available.

### CODE REVIEW

**Review by aaronallen8455 — COMMENTED** [review]
Definitely an improvement on Poppers.

I noticed you got rid of the query caching stuff - seems like a good idea.

I found a lot of typos in the haddocks so I made branch aallen/fix-plan-haddock-typos to fix them rather than pointing them out individually.

**Review by telser — APPROVED** [review]

File: orville-postgresql/src/Database/Orville/PostgreSQL/Plan/Operation.hs [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+  , askParam
+  , assertRight
+
+  , SelectOperation(..)
+  , selectOperation
+  ) where
+
+import qualified Control.Monad.Catch as Catch
+import qualified Data.Maybe as Maybe
+import qualified Data.Map as Map
```
[the remark this answers is not in the document]
Comment (qxjit) [reply orville-postgresql/src/Database/Orville/PostgreSQL/Plan/Operation.hs]: Done

File: orville-postgresql/src/Database/Orville/PostgreSQL/Plan/Operation.hs [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+      $ params
+
+    insertRow row results =
+      Map.alter
+        (\mbRows -> Just (row : Maybe.fromMaybe [] mbRows))
+        (categorizeRow selectOp row)
+        results
+
+    rowMap =
+      produceResult selectOp <$> foldr insertRow emptyRowsMap rows
```
[the remark this answers is not in the document]
Comment (qxjit) [reply orville-postgresql/src/Database/Orville/PostgreSQL/Plan/Operation.hs]: Done

File: orville-postgresql/src/Database/Orville/PostgreSQL/Plan/Operation.hs [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+  rows <- Select.runSelect . selectMany selectOp $ params
+
+  let
+    -- Seed add initial map with an empty list for every input parameter
+    -- to guarantee that each param is a key in the map even if no rows
+    -- where returned from the select query for that param.
+    emptyRowsMap =
+      Map.fromList
+      . map (\param -> (param, []))
+      $ params
```
[the remark this answers is not in the document]
Comment (qxjit) [reply orville-postgresql/src/Database/Orville/PostgreSQL/Plan/Operation.hs:392]: To be honest, I prefer the lambda over the combination of `zip` and `repeat`.

File: orville-postgresql/src/Database/Orville/PostgreSQL/Plan.hs [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+  when the plan is executed.
+
+  The 'scope' type is used internally by Orville to track the plan is currently
+  executed against a single input or multiple inputs. This type parameter
+  should never specified as a concrete type in user code, but must be exposed
+  as a variable to ensure that execute scope is tracked correctly through
+  usages of 'bind'.
+
+-}
+data Plan scope param result where
```
[the remark this answers is not in the document]
Comment (qxjit) [reply orville-postgresql/src/Database/Orville/PostgreSQL/Plan.hs:69]: I was able to introduce an explicit `ManyScope` tag type that replaces using `Many k` as the to change this to just `Type`

File: orville-postgresql/src/Database/Orville/PostgreSQL/Plan.hs [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+      pure (Many.apply manyFs manyAs)
+
+    Chain planAB planBC -> do
+      bs <- executeMany planAB params
+      cs <- executeMany planBC (Many.elems bs)
+      pure $ Many.compose cs bs
+
+data Explain a where
+  ExplainOne  :: forall b. Explain b
+  ExplainMany :: forall b. Explain b
```
[the remark this answers is not in the document]
Comment (qxjit) [reply orville-postgresql/src/Database/Orville/PostgreSQL/Plan.hs]: This ended up being address by introducing `ManyScope`

**Comment by qxjit** [comment]
@aaronallen8455 :heart: I merged you branch in

**Review by aaronallen8455 — APPROVED** [review]

**Pull request merged**